README.md
# statsearchanalyticsr

The statsearchanalyticsr makes it easy to pull [STAT Search Analytics](https://getstat.com) data into R for visualization, integration, and analysis by enabling you to retrieve keyword ranking data in a tidy way. 

## Installation

```
# Install from CRAN:
install.packages("statsearchanalyticsr")

# Or the development version from GitHub:
install.packages('remotes')
remotes::install_github('searchdiscovery/statsearchanalyticsr')
```

## Define global variables with account information

Add both of these variables to your .Renviron file and restart the R session.

```r
SSAR_APIKEY = '{apikey}'  
SSAR_SUBDOMAIN = '{subdomain}'
```
Once completed, then the API is ready to start pulling data.

### Getting an API Key

The API Key can be acquired following these steps:

1. Under the Options menu, click Account Management.
2. In the Account Management pop-up, select Account Settings.
3. Click on the New API Key button.
4. In the pop-up, click Yes to generate a new API key and deactivate the old one.

*[More information.](https://help.getstat.com/knowledgebase/api-services/#generating-an-api-key)*

### Getting the subdomain

The subdomain is easy to find. It is the first part of the 'getstat.com' URL. For example, if your STAT login URL is `demonstration.getstat.com` then your STAT subdomain is `demonstration`.

*** 
### Rate Limits

The default daily limit for API calls is 1,000 calls per day. The daily count resets every night at midnight UTC. If you need to exceed the daily limit, you will need to talk to STAT Searh Analytics support.

If you are attempting to pull more than 1,000 keywords at a time, then consider using the ssar_bulk_rankings() function. This function enables you to pull all the rankings by day for all keywords. This option can be more time efficient, saving you a 24 hour wait for next data pull, but note it will likely include much more data than necessary.

## Examples

This is a basic example showing the process of pulling data for the first time:

``` r
## Get all available projects
projects <- ssar_projects()

## Get the sites for a specific project
sites <- ssar_sites(projectid = {project_id}) #replace {project_id} with a specific project id found in the `projects` object

## Get the keywords for a site
keywords <- ssar_keywords(siteid = {site_id}) #replace {site_id} with a specific site id found in the 'sites' object

## Get the rankings for a keyword
rankings <- ssar_rankings(keywordid = {keyword_id}, #replace {keyword_id} with a specific keyword id  found in the `keywords` object
                          fromdate = {created_at}, 
                          todate = Sys.Date()-1) 

## Get the bulk rankings report
all_rankings <- ssar_bulk_rankings(date = Sys.Date()-2, 
                                  siteid = {site_id}, 
                                  ranktype = 'highest', 
                                  engines = 'google', 
                                  currentlytracked = 'true', 
                                  crawledkeywords = 'true')
```



NEWS.md
# statsearchanalyticsr (development version)

# statsearchanalyticsr 0.1.4

* Initial CRAN version

# statsearchanalyticsr 0.1.3

* Changed the bulk ranking request to message the correct number of keywords in the result

# statsearchanalyticsr 0.1.2

* Added Bulk Reports function
* Added  Sites Ranking Distribution function
* Minor documentation improvements

# statsearchanalyticsr 0.1.1

* Improved Bulk Rankings report function to help mitigate the 1000 api calls per day limit. Now you can pull in all keywords by day and only do one api call for the day, no matter how many keywords your account has.

# statsearchanalyticsr 0.1.0

* Added a `NEWS.md` file to track changes to the package.
* Added bulk rankings and request option


R/get_sites.R
#' Get Sites
#'
#' Retrieve a table of all the sites and metadata in a specified project
#'
#' @param projectid The project id. If not provided then all sites will be returned.
#' @param start If more than `results` are available use `start` as pagination. Index starts at 0 (default).
#' @param results Default is 100. Max is 5000.
#' @param subdomain The account subdomain
#' @param apikey The api key from the account
#'
#' @return A table of the site information within a project
#' 
#' @examples 
#' \dontrun{
#' ssar_sites(projectid = {project_id}, #replace with your project id
#'            results = 300)
#' }
#'
#' @import httr tidyr jsonlite
#' @importFrom glue glue
#' @importFrom purrr map_df
#'
#' @export
#'
ssar_sites <- function( projectid = NULL,
                         start = NULL,
                         results = 100,
                         subdomain = Sys.getenv('SSAR_SUBDOMAIN'),
                         apikey = Sys.getenv('SSAR_APIKEY')) {
#add valid params
  params <- list(project_id = projectid, start = start, results = results, format = 'json')
#collec non NULL params into a list
  valid_params <- Filter(Negate(is.null), params)
#make that list into a parameter string
  urlparams <- paste0(names(valid_params),'=',valid_params, collapse = '&')
  # build request url
  baseurl <- glue::glue('https://{subdomain}.getstat.com/api/v2/{apikey}/')

  endpoint <- if(!is.null(projectid)){
    'sites/list'
  } else {
    'sites/all'
  }

  requrl <- glue::glue('{baseurl}{endpoint}?{urlparams}&format=json')

 sites <- httr::GET(requrl)

#check the status for the call and return errors or don't
 httr::stop_for_status(sites, glue::glue('get the sites list. \n {httr::content(sites)$Result}'))
 #if 200 but no results due to an error
 if(is.null(httr::content(sites)$Response)) {
   stop(httr::content(sites)$Result)
 }

 #return the results
 sites <- httr::content(sites)[[1]]$Result

 df <-  purrr::map_df(sites, unlist)

return(df)
}


R/get_keywords.R
#' Get Keywords
#'
#' Retrieve a table of all the keywords in a particular site and the corresponding metadata
#'
#' @param siteid The site id (required)
#' @param start The default is 0 (zero indexed)
#' @param results The default is 100
#' @param subdomain The account subdomain
#' @param apikey The api key from the account
#' 
#' @examples  
#' \dontrun{
#' ssar_keywords(siteid = {site_id} ) #replace is your site id
#' }
#'
#' @import httr tidyr jsonlite
#' @importFrom glue glue
#' @importFrom purrr map_df
#'
#' @export
#' 
#' @return A dataframe of keywords along with 29 other columns of data
#'
ssar_keywords <- function(siteid = NULL,
                         start = 0,
                         results = 100,
                         subdomain = Sys.getenv('SSAR_SUBDOMAIN'),
                         apikey = Sys.getenv('SSAR_APIKEY')) {
#siteid check
  if(is.null(siteid)) stop ("The argument 'siteid' is required. Please include 'siteid' as an argument.")

  #add valid params
  params <- list(site_id = siteid, start = start, results = results, format = 'json')
  #collec non NULL params into a list
  valid_params <- Filter(Negate(is.null), params)
  #make that list into a parameter string
  urlparams <- paste0(names(valid_params),'=',valid_params, collapse = '&')
  # build request url
  baseurl <- glue::glue('https://{subdomain}.getstat.com/api/v2/{apikey}/')
  endpoint <- 'keywords/list'

  requrl <- glue::glue('{baseurl}{endpoint}?{urlparams}')


 keywords <- httr::GET(requrl)

 #check the status for the call and return errors if found
 httr::stop_for_status(keywords, glue::glue('get the keywords. \n {httr::content(keywords)$Result}'))
 #if 200 but no results due to an error
 if(is.null(httr::content(keywords)$Response)) {
   stop(httr::content(keywords)$Result)
 }

 keyw <- httr::content(keywords)$Response$Result

 df <-  purrr::map_df(keyw, unlist)

 return(df)

}


R/get_projects.R
#' Get Projects
#'
#' Receive a list of all the projects on an account accessible through the provided api key
#'
#' @param subdomain The account subdomain
#' @param apikey The api key from the account
#'
#' @import httr tidyr jsonlite
#' @importFrom glue glue
#' @importFrom purrr map_df
#'
#' @return A dataframe of available project data your authentication has access to
#' 
#' @examples
#' \dontrun{
#' projects(subdomain = Sys.getenv('SSAR_SUBDOMAIN'),
#'          apikey = Sys.getenv('SSAR_APIKEY'))
#' }
#' 
#' @export
#' 
#'
ssar_projects <- function(subdomain = Sys.getenv('SSAR_SUBDOMAIN'),
                            apikey = Sys.getenv('SSAR_APIKEY')) {

# add valid params
  params <- list(format = 'json')

#collec non NULL params into a list
  valid_params <- Filter(Negate(is.null), params)

#make that list into a parameter string
  urlparams <- paste0(names(valid_params),'=',valid_params, collapse = '&')
# build request url
  baseurl <- glue::glue('https://{subdomain}.getstat.com/api/v2/{apikey}/')
  endpoint <- 'projects/list'

  requrl <- glue::glue('{baseurl}{endpoint}?{urlparams}&format=json')

  projects <- httr::GET(requrl)

  #check the status for the call and return errors if found
  httr::stop_for_status(projects, glue::glue('get the projects. \n {httr::content(projects)$Result}'))
  #if 200 but no results due to an error
  if(is.null(httr::content(projects)$Response)) {
    stop(httr::content(projects)$Result)
  }

 projs <- httr::content(projects)[[1]]$Result

 df <-  purrr::map_df(projs, unlist)

 return(df)

}


R/get_bulk_request.R
#' Request Bulk Report
#'
#' Retrieve a bulk report using a bulk report id. Intended to be used with the `get_bulk_rankings` function but can be used independently as well.
#'
#' @param id Id of the bulk job #required
#' @param subdomain The account subdomain
#' @param apikey The api key from the account
#'
#' @return list of the bulk keyword ranking report for the requested ID
#' 
#' @examples 
#' \dontrun{
#' ssar_bulk_request(id = {report_id})
#' }
#'
#' @import httr tidyr jsonlite
#' @importFrom glue glue
#' @importFrom purrr map_df
#' @importFrom stringr str_c
#' 
#' @export
#'
#'
ssar_bulk_request <- function(id = NULL,
                               subdomain = Sys.getenv('SSAR_SUBDOMAIN'),
                               apikey = Sys.getenv('SSAR_APIKEY')) {
if(is.null(id)){
  stop("id argument cannot be empty")
}
  # build request url
  baseurl <- glue::glue('https://{subdomain}.getstat.com/api/v2/{apikey}/')

  endpoint <- 'bulk/status'

  requrl <- glue::glue('{baseurl}{endpoint}?id={id}&format=json')

 bulk_status <- httr::GET(requrl)
#check the status for the call and return errors or don't
 httr::stop_for_status(bulk_status, glue::glue('get the bulk rankings. \n {httr::content(bulkranks)$Result}'))
 #if 200 but no results due to an error
 if(is.null(httr::content(bulk_status)$Response)) {
   stop(httr::content(bulk_status)$Result)
 }

 #return the results
 status <- httr::content(bulk_status)$Response$Result$Status

 if(status == 'Completed') {
 data <- httr::GET(httr::content(bulk_status)$Response$Result$StreamUrl)

 data_content <- httr::content(data)$Response$Project

 return(data_content)
 } else {
  return(status)
 }
}


R/get_bulk_rankings.R
#' Get Bulk Rankings Report
#'
#' Retrieve a bulk report of all the rankings or for specific sites.
#'
#' @param date The date being requested (required) in 'YYYY-MM-DD' format. Default is set to yesterday.
#' @param siteid The site id. If not provided then all sites will be returned. Comma separated list of specific site IDs, default is all sites.
#' @param ranktype This argument changes the call between getting the highest ranks for the keywords for the date with the value highest, or getting all the ranks for each engine for a keyword for a date with the value all. Defaults to highest if not provided.
#' @param engines This argument lets you choose which search engines to include in the export, defaulting to Google and Bing. Engines can be passed in comma separated to get multiple.
#' @param currentlytracked This argument will cause the API to ouptput only keywords which currently have tracking on at the time the API request is generated.
#' @param crawledkeywords This argument causes the API to only include output for keywords that were crawled on the date argument provided.
#' @param subdomain The account subdomain
#' @param apikey The api key from the account
#'
#' @return The dataframe with all keywords ranking information for the requested date
#'
#' @import httr tidyr jsonlite
#' @importFrom glue glue
#' @importFrom purrr map_df
#' @importFrom stringr str_c
#'
#' @export
#'
ssar_bulk_rankings <- function( date = Sys.Date()-1,
                                siteid = NULL,
                                ranktype = 'highest',
                                engines = c('google', 'bing'),
                                currentlytracked = TRUE,
                                crawledkeywords = TRUE,
                                subdomain = Sys.getenv('SSAR_SUBDOMAIN'),
                                apikey = Sys.getenv('SSAR_APIKEY')) {
  if(is.null(date)){
    stop("date argument cannot be empty")
  }
  if(length(date) > 1){
    stop("date argument cannot include more than one day")
  }
  #clean engines argument
  if(length(engines) > 1) {
    engines <- stringr::str_c(engines, collapse = ',')
  }
  #clean sites
  if(length(siteid) > 1) {
    siteid <- stringr::str_c(siteid, collapse = ',')
  }
  #add valid params
  params <- list(site_id = siteid, rank_type = ranktype, engines = engines, currently_tracked_only=currentlytracked, crawled_keywords_only = crawledkeywords, format = 'json')
  
  #collec non NULL params into a list
  valid_params <- Filter(Negate(is.null), params)
  #make that list into a parameter string
  urlparams <- paste0(names(valid_params),'=',valid_params, collapse = '&')
  # build request url
  baseurl <- glue::glue('https://{subdomain}.getstat.com/api/v2/{apikey}/')
  
  endpoint <- 'bulk/ranks'
  
  requrl <- glue::glue('{baseurl}{endpoint}?date={as.Date(date)}&{urlparams}&format=json')
  
  bulk_ranks <- httr::GET(requrl)
  
  #check the status for the call and return errors or don't
  httr::stop_for_status(bulk_ranks, glue::glue('get the bulk rankings. \n   {httr::content(bulk_ranks)$Result}'))
  #if 200 but no results due to an error
  if(is.null(httr::content(bulk_ranks)$Response)) {
    stop(httr::content(bulk_ranks)$Result)
  }
  
  #return the results
  report_id <- httr::content(bulk_ranks)$Response$Result$Id
  
  message(glue::glue('Your report id is {report_id} \n Checking to see if it is complete...'))
  
  requrl <- glue::glue('{baseurl}bulk/status?id={report_id}&format=json')
  
  statusgot <- httr::GET(requrl)
  
  
  try = 1
  ##Check if the report is ready
  checkit <- function(statusgot, try){
    keepgoingstatus <- list('NotStarted','InProgress')
    stopstatus <- list('Deleted','Failed')
    completestatus <- list('Completed')
    res <- httr::content(statusgot)
    status <- httr::content(statusgot)$Response$Result$Status
    
    if(status %in% keepgoingstatus) {
      message(glue::glue('Attempt #{try}'))
      message(glue::glue('The current status is {status}.'))
      
      #pause for 2 seconds for the report to run
      Sys.sleep(2)
      
      #get the status for the report
      status_res <- httr::GET(requrl) 
      try <- try + 1
      checkit(status_res, try)
    } else if(status %in% stopstatus) {
      #report back the status and ask to retry
      message(glue::glue('The current status of the report is \'{status}\'. You will need to retry your request.'))
    } else if(status %in% completestatus) {
      
      message(glue::glue('Good news! Your report is complete'))
      report <- httr::GET(res$Response$Result$StreamUrl)
      
      reportinfo <- httr::content(report)
      
      resultkeywords <- purrr::map_df(reportinfo$Response$Project$Site$Keyword, unlist)
      #message the number of keywords that were returned.
      message(glue::glue('{length(unique(resultkeywords$Id))} Total keywords have been returned for {date}'))
      
      #return the final keyword report for the date requested
      return(resultkeywords)
    }
  }
  report_response <- checkit(statusgot, try)
  
}


R/get_sites_ranking_distribution.R
#' Get Sites Ranking Distribution
#'
#' This function returns all ranking distribution records for Google and Bing for a site with the specified id. The maximum date range can be no greater than 31 days.
#'
#' @param siteid *Required* The site id.
#' @param fromdate *Required* Character string in the 'YYYY-MM-DD' format. Default is -31 days from today since the maximum date range can be no greater than 31 days.
#' @param todate *Required* Character string in the 'YYYY-MM-DD' format. Default is yesterday.
#' @param subdomain The account subdomain
#' @param apikey The api key from the account
#'
#' @return A table with Google, GoogleBaseRank, and Bing ranking distributions by date
#' 
#' @examples  
#' \dontrun{
#' ssar_sites_ranking_dist(siteid = {site_id}, #replace is your site id
#'                        fromdate = '2021-04-01',
#'                        todate = '2021-05-31')
#' }
#' 
#' @import httr tidyr jsonlite
#' @importFrom glue glue
#' @importFrom lubridate date
#' @importFrom purrr map_df
#'
#' @export
#'
ssar_sites_ranking_dist <- function( siteid = NULL,
                        fromdate = as.character(Sys.Date()-31),
                        todate = as.character(Sys.Date()-1),
                        subdomain = Sys.getenv('SSAR_SUBDOMAIN'),
                        apikey = Sys.getenv('SSAR_APIKEY') ) {
  
#siteid must be included
  if(is.null(siteid)) {
    stop('\'siteid\' must be provided')
  }
#add valid params
  params <- list(id = siteid, from_date = fromdate, to_date = todate, format = 'json')
#collec non NULL params into a list
  valid_params <- Filter(Negate(is.null), params)
#make that list into a parameter string
  urlparams <- paste0(names(valid_params),'=',valid_params, collapse = '&')
  # build request url
  baseurl <- glue::glue('https://{subdomain}.getstat.com/api/v2/{apikey}/')

  endpoint <- 'sites/ranking_distributions'

  requrl <- glue::glue('{baseurl}{endpoint}?{urlparams}&format=json')

 sites_res <- httr::GET(requrl)

#check the status for the call and return errors or don't
 httr::stop_for_status(sites_res, glue::glue('get the sites list. \n {httr::content(sites_res)$Result}'))
 #if 200 but no results due to an error
 if(is.null(httr::content(sites_res)$Response$RankDistribution)) {
   stop("The request did not return any results. Make sure you used the correct Site ID.")
 }

 #return the results
 sites <- httr::content(sites_res)$Response$RankDistribution

 dists <- function(x){
   google <- tidyr::as_tibble(c(date = sites[[x]]$date, type ='Google', sites[[x]]$Google ))
   googlebaserank <- tidyr::as_tibble(c(date = sites[[x]]$date, type ='GoogleBaseRank', sites[[x]]$GoogleBaseRank))
   bing <- tidyr::as_tibble(c(date = sites[[x]]$date, type ='Bing', sites[[x]]$Bing))
   
   rbind(google, googlebaserank, bing)
 }
 
days = as.numeric(lubridate::date(todate)-lubridate::date(fromdate))
df <- purrr::map_df(seq(days), dists)
 
return(df)
}


R/get_rankings.R
#' Get Rankings
#'
#' Retrieve a table including metadata for all the sites in a particular project
#'
#' @param keywordid The keyword id (required)
#' @param fromdate Default is 100 most recent results (optional) Format is a string YYYY-MM-DD
#' @param todate Default is most recent ranking day (optional) Format is a string YYYY-MM-DD
#' @param start The default is 0 (zero indexed). The starting result for paginated requests
#' @param results The default is 100
#' @param subdomain The account subdomain
#' @param apikey The api key from the account
#'
#' @return A dataframe of all rankings within a defined date range for a specific keyword
#' 
#' @examples 
#' \dontrun{
#' ssar_rankings(keywordid = {keyword_id}, #replace with your keyword_id 
#'               fromdate = '2021-01-01', 
#'               todate = '2021-04-01')
#' }
#'
#' @import httr tidyr jsonlite
#' @importFrom glue glue
#' @importFrom purrr map_df
#'
#' @export
#'
ssar_rankings <- function(keywordid = NULL, #required
                         fromdate = NULL,
                         todate = NULL,
                         start = 0,
                         results = 100,
                         subdomain = Sys.getenv('SSAR_SUBDOMAIN'),
                         apikey = Sys.getenv('SSAR_APIKEY')) {
#siteid check
  if(is.null(keywordid)) stop ("The argument 'keywordid' is required. Please include 'keywordid' as an argument.")

  #add valid params
  params <- list(keyword_id = keywordid, from_date = fromdate, to_date = todate, start = start, results = results, format = 'json')
  #collect non NULL params into a list
  valid_params <- Filter(Negate(is.null), params)
  #make that list into a parameter string
  urlparams <- paste0(names(valid_params),'=',valid_params, collapse = '&')
  # build request url
  baseurl <- glue::glue('https://{subdomain}.getstat.com/api/v2/{apikey}/')
  endpoint <- 'rankings/list'

  requrl <- glue::glue('{baseurl}{endpoint}?{urlparams}')


 rankings <- httr::GET(requrl)

 #check the status for the call and return errors or don'te
 httr::stop_for_status(rankings, glue::glue('get the rankings. \n {httr::content(rankings)$Result}'))
  #if 200 but no results due to an error
 if(is.null(httr::content(rankings)$Response)) {
  stop(httr::content(rankings)$Result)
}

 rank <- httr::content(rankings)$Response$Result

 df <-  purrr::map_df(rank, unlist)

 return(df)

}
